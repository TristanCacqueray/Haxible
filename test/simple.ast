[ Definition
    { name = "playbook"
    , requires = []
    , provides = []
    , outputs = Environment
        { getEnv = [] }
    , exprs =
        [ Expr
            { binder = Binder "resultsLocalhost0"
            , requires = []
            , provides =
                [ Resource
                    { name = Binder "stat0"
                    , dep = Register "etc"
                    }
                , Resource
                    { name = Binder "stat0"
                    , dep = Path "/etc"
                    }
                ]
            , outputs = Right []
            , requirements = []
            , loop = Nothing
            , term = DefinitionCall
                ( CallDefinition
                    { name = "playLocalhost0"
                    , playAttrs =
                        [
                            ( "hosts"
                            , String "localhost"
                            )
                        ]
                    , baseEnv = []
                    }
                )
            }
        , Expr
            { binder = Binder "resultsZuulExecutor0"
            , requires = []
            , provides =
                [ Resource
                    { name = Binder "file0"
                    , dep = Register "zuuldir"
                    }
                , Resource
                    { name = Binder "file0"
                    , dep = Path "/tmp/zuul"
                    }
                ]
            , outputs = Right []
            , requirements = []
            , loop = Nothing
            , term = DefinitionCall
                ( CallDefinition
                    { name = "playZuulExecutor0"
                    , playAttrs =
                        [
                            ( "hosts"
                            , String "zuul_executor"
                            )
                        ]
                    , baseEnv = []
                    }
                )
            }
        , Expr
            { binder = Binder "resultsNodepoolLauncher0"
            , requires = []
            , provides =
                [ Resource
                    { name = Binder "file1"
                    , dep = Path "/tmp/nodepool"
                    }
                ]
            , outputs = Right []
            , requirements = []
            , loop = Nothing
            , term = DefinitionCall
                ( CallDefinition
                    { name = "playNodepoolLauncher0"
                    , playAttrs =
                        [
                            ( "hosts"
                            , String "nodepool_launcher"
                            )
                        ]
                    , baseEnv = []
                    }
                )
            }
        , Expr
            { binder = Binder "resultsLocalhost1"
            , requires =
                [ Resource
                    { name = Binder "stat0"
                    , dep = Register "etc"
                    }
                ]
            , provides = []
            , outputs = Right []
            , requirements =
                [ Requirement
                    { name = "etc"
                    , origin = Nested
                        ( Binder "resultsLocalhost0" ) 0
                    }
                ]
            , loop = Nothing
            , term = DefinitionCall
                ( CallDefinition
                    { name = "playLocalhost1"
                    , playAttrs =
                        [
                            ( "hosts"
                            , String "localhost"
                            )
                        ]
                    , baseEnv = []
                    }
                )
            }
        ]
    }
, Definition
    { name = "playLocalhost0"
    , requires = []
    , provides =
        [ Resource
            { name = Binder "stat0"
            , dep = Register "etc"
            }
        , Resource
            { name = Binder "stat0"
            , dep = Path "/etc"
            }
        ]
    , outputs = Environment
        { getEnv =
            [
                ( Binder "stat0"
                , Right
                    [ Resource
                        { name = Binder "stat0"
                        , dep = Register "etc"
                        }
                    , Resource
                        { name = Binder "stat0"
                        , dep = Path "/etc"
                        }
                    ]
                )
            ,
                ( Binder "command0"
                , Right []
                )
            ]
        }
    , exprs =
        [ Expr
            { binder = Binder "stat0"
            , requires = []
            , provides =
                [ Resource
                    { name = Binder "stat0"
                    , dep = Register "etc"
                    }
                , Resource
                    { name = Binder "stat0"
                    , dep = Path "/etc"
                    }
                ]
            , outputs = Right
                [ Resource
                    { name = Binder "stat0"
                    , dep = Register "etc"
                    }
                , Resource
                    { name = Binder "stat0"
                    , dep = Path "/etc"
                    }
                ]
            , requirements = []
            , loop = Nothing
            , term = ModuleCall
                ( CallModule
                    { module_ = "stat"
                    , params = Object
                        ( fromList
                            [
                                ( "path"
                                , String "/etc"
                                )
                            ]
                        )
                    , taskAttrs = []
                    }
                )
            }
        , Expr
            { binder = Binder "command0"
            , requires =
                [ Resource
                    { name = Binder "stat0"
                    , dep = Register "etc"
                    }
                ]
            , provides = []
            , outputs = Right []
            , requirements =
                [ Requirement
                    { name = "etc"
                    , origin = Direct
                        ( Binder "stat0" )
                    }
                , Requirement
                    { name = "_etc"
                    , origin = Direct
                        ( Binder "stat0" )
                    }
                ]
            , loop = Nothing
            , term = ModuleCall
                ( CallModule
                    { module_ = "command"
                    , params = String "echo etc exist: {{ etc.stat.exists }}"
                    , taskAttrs = []
                    }
                )
            }
        ]
    }
, Definition
    { name = "playZuulExecutor0"
    , requires = []
    , provides =
        [ Resource
            { name = Binder "file0"
            , dep = Register "zuuldir"
            }
        , Resource
            { name = Binder "file0"
            , dep = Path "/tmp/zuul"
            }
        ]
    , outputs = Environment
        { getEnv =
            [
                ( Binder "file0"
                , Right
                    [ Resource
                        { name = Binder "file0"
                        , dep = Register "zuuldir"
                        }
                    , Resource
                        { name = Binder "file0"
                        , dep = Path "/tmp/zuul"
                        }
                    ]
                )
            ,
                ( Binder "command1"
                , Right []
                )
            ]
        }
    , exprs =
        [ Expr
            { binder = Binder "file0"
            , requires = []
            , provides =
                [ Resource
                    { name = Binder "file0"
                    , dep = Register "zuuldir"
                    }
                , Resource
                    { name = Binder "file0"
                    , dep = Path "/tmp/zuul"
                    }
                ]
            , outputs = Right
                [ Resource
                    { name = Binder "file0"
                    , dep = Register "zuuldir"
                    }
                , Resource
                    { name = Binder "file0"
                    , dep = Path "/tmp/zuul"
                    }
                ]
            , requirements = []
            , loop = Nothing
            , term = ModuleCall
                ( CallModule
                    { module_ = "file"
                    , params = Object
                        ( fromList
                            [
                                ( "path"
                                , String "/tmp/zuul"
                                )
                            ,
                                ( "state"
                                , String "directory"
                                )
                            ]
                        )
                    , taskAttrs = []
                    }
                )
            }
        , Expr
            { binder = Binder "command1"
            , requires =
                [ Resource
                    { name = Binder "file0"
                    , dep = Path "/tmp/zuul"
                    }
                ]
            , provides = []
            , outputs = Right []
            , requirements =
                [ Requirement
                    { name = "zuuldir"
                    , origin = Direct
                        ( Binder "file0" )
                    }
                , Requirement
                    { name = "_tmp_zuul"
                    , origin = Direct
                        ( Binder "file0" )
                    }
                ]
            , loop = Nothing
            , term = ModuleCall
                ( CallModule
                    { module_ = "command"
                    , params = String "echo Starting executor -d /tmp/zuul"
                    , taskAttrs = []
                    }
                )
            }
        ]
    }
, Definition
    { name = "playNodepoolLauncher0"
    , requires = []
    , provides =
        [ Resource
            { name = Binder "file1"
            , dep = Path "/tmp/nodepool"
            }
        ]
    , outputs = Environment
        { getEnv =
            [
                ( Binder "file1"
                , Right
                    [ Resource
                        { name = Binder "file1"
                        , dep = Path "/tmp/nodepool"
                        }
                    ]
                )
            ,
                ( Binder "command2"
                , Right []
                )
            ,
                ( Binder "debug0"
                , Right []
                )
            ]
        }
    , exprs =
        [ Expr
            { binder = Binder "file1"
            , requires = []
            , provides =
                [ Resource
                    { name = Binder "file1"
                    , dep = Path "/tmp/nodepool"
                    }
                ]
            , outputs = Right
                [ Resource
                    { name = Binder "file1"
                    , dep = Path "/tmp/nodepool"
                    }
                ]
            , requirements = []
            , loop = Nothing
            , term = ModuleCall
                ( CallModule
                    { module_ = "file"
                    , params = Object
                        ( fromList
                            [
                                ( "path"
                                , String "/tmp/nodepool"
                                )
                            ,
                                ( "state"
                                , String "directory"
                                )
                            ]
                        )
                    , taskAttrs = []
                    }
                )
            }
        , Expr
            { binder = Binder "command2"
            , requires =
                [ Resource
                    { name = Binder "file1"
                    , dep = Path "/tmp/nodepool"
                    }
                ]
            , provides = []
            , outputs = Right []
            , requirements =
                [ Requirement
                    { name = "_tmp_nodepool"
                    , origin = Direct
                        ( Binder "file1" )
                    }
                ]
            , loop = Nothing
            , term = ModuleCall
                ( CallModule
                    { module_ = "command"
                    , params = String "echo Starting scheduler -d /tmp/nodepool"
                    , taskAttrs = []
                    }
                )
            }
        , Expr
            { binder = Binder "debug0"
            , requires = []
            , provides = []
            , outputs = Right []
            , requirements = []
            , loop = Nothing
            , term = ModuleCall
                ( CallModule
                    { module_ = "debug"
                    , params = Null
                    , taskAttrs = []
                    }
                )
            }
        ]
    }
, Definition
    { name = "playLocalhost1"
    , requires =
        [ Resource
            { name = Binder "stat0"
            , dep = Register "etc"
            }
        ]
    , provides = []
    , outputs = Environment
        { getEnv =
            [
                ( Binder "debug1"
                , Right []
                )
            ]
        }
    , exprs =
        [ Expr
            { binder = Binder "debug1"
            , requires =
                [ Resource
                    { name = Binder "stat0"
                    , dep = Register "etc"
                    }
                ]
            , provides = []
            , outputs = Right []
            , requirements = []
            , loop = Nothing
            , term = ModuleCall
                ( CallModule
                    { module_ = "debug"
                    , params = Object
                        ( fromList
                            [
                                ( "msg"
                                , String "etc stats is {{ etc }}"
                                )
                            ]
                        )
                    , taskAttrs = []
                    }
                )
            }
        ]
    }
]