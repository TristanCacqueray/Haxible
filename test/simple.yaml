- hosts: localhost
  tasks:
    - stat:
        path: /tmp
      register: tmpstat

    - command: "echo tmp exist: {{ tmpstat.stat.exists }}"
