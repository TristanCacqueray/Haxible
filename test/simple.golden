( Playbook
    [ HostPlay
        { hosts = "localhost"
        , tasks =
            [ Task
                { name = Nothing
                , action = "stat"
                , attributes = Object
                    ( fromList
                        [
                            ( "path"
                            , String "/tmp"
                            )
                        ]
                    )
                , requires = []
                , register = Just "tmpstat"
                , loop = Nothing
                , vars = Nothing
                }
            , Task
                { name = Nothing
                , action = "command"
                , attributes = Object
                    ( fromList
                        [
                            ( "_raw_params"
                            , String "echo tmp exist: {{ tmpstat.stat.exists }}"
                            )
                        ]
                    )
                , requires = [ "tmpstat" ]
                , register = Nothing
                , loop = Nothing
                , vars = Nothing
                }
            ]
        , vars = Nothing
        }
    ]
, "{-# LANGUAGE QuasiQuotes, ApplicativeDo, OverloadedStrings #-}
  {- cabal:
  build-depends: base, haxible
  ghc-options: -threaded -rtsopts -with-rtsopts=-N -with-rtsopts=-T
  -}
  module Main (main) where

  import Haxible.Eval

  main :: IO ()
  main = runHaxible playbook

  playbook :: AnsibleHaxl ()
  playbook = do
    play_localhost
    pure ()

  play_localhost :: AnsibleHaxl ()
  play_localhost = do
    tmpstat <- runTask "localhost" (Nothing) "stat" [json| {"path":"/tmp"} |]
    runTask "localhost" (Nothing) "command" (renderTemplates [("tmpstat", tmpstat)] [json| {"_raw_params":"echo tmp exist: {{ tmpstat.stat.exists }}"} |])
    pure ()
  "
)